vector/src/main/java/im/vector/app/features/onboarding/RegistrationActionHandler.kt

```diff
@@ -16,105 +16,93 @@
-                        else -> RegistrationResult.Error(it)
-                    }
-                }
-        )
+    private suspend fun processFlowResult(result: RegistrationResult.NextStep, state: SelectedHomeserverState): Result {
+        return if (shouldFastTrackDummyAction(result)) {
+            processAction(state, RegisterAction.RegisterDummy)
+        } else {
+            handleNextStep(state, result.flowResult)
+        }
     }
 
-    private tailrec suspend fun handleCheckIfEmailIsValidated(registrationWizard: RegistrationWizard, delayMillis: Long): RegistrationResult {
-        return runCatching { registrationWizard.checkIfEmailHasBeenValidated(delayMillis) }.fold(
-                onSuccess = { it.toRegistrationResult() },
-                onFailure = {
-                    when {
-                        it.is401() -> null // recursively continue to check with a delay
-                        else -> RegistrationResult.Error(it)
-                    }
-                }
-        ) ?: handleCheckIfEmailIsValidated(registrationWizard, 10_000)
-    }
-}
+    private fun shouldFastTrackDummyAction(result: RegistrationResult.NextStep) = authenticationService.isRegistrationStarted() &&
+            result.flowResult.missingStages.hasMandatoryDummy()
 
-private inline fun resultOf(block: () -> MatrixRegistrationResult): RegistrationResult {
-    return runCatching { block() }.fold(
-            onSuccess = { it.toRegistrationResult() },
-            onFailure = { RegistrationResult.Error(it) }
-    )
-}
-
-private fun MatrixRegistrationResult.toRegistrationResult() = when (this) {
-    is FlowResponse -> RegistrationResult.NextStep(flowResult)
-    is Success -> RegistrationResult.Complete(session)
-}
-
-sealed interface RegistrationResult {
-    data class Error(val cause: Throwable) : RegistrationResult
-    data class Complete(val session: Session) : RegistrationResult
-    data class NextStep(val flowResult: FlowResult) : RegistrationResult
-    data class SendEmailSuccess(val email: String) : RegistrationResult
-}
-
-sealed interface RegisterAction {
-    object StartRegistration : RegisterAction
-    data class CreateAccount(val username: String, val password: String, val initialDeviceName: String) : RegisterAction
+    private suspend fun handleNextStep(state: SelectedHomeserverState, flowResult: FlowResult): Result {
+        return when {
+            flowResult.registrationShouldFallback() -> Result.UnsupportedStage
+            authenticationService.isRegistrationStarted() -> findNextStage(state, flowResult)
+            else -> Result.StartRegistration
+        }
+    }
 
-    data class AddThreePid(val threePid: RegisterThreePid) : RegisterAction
-    object SendAgainThreePid : RegisterAction
+    private fun findNextStage(state: SelectedHomeserverState, flowResult: FlowResult): Result {
+        val orderedResult = when {
```

Comment:
I think `if` would be much more readable here. `when` doesn't looks good with complex conditions. But I won't insist